### ISSUE TITLE
how does the speed compare with opencv?

### ISSUE THREAD

**Issue by jinfagang** [Cykooz/fast_image_resize#12]
how does the speed compare with opencv?

**Comment by Cykooz** [comment]
I have not yet done a benchmark to measure the speed of opencv. 
I plan to do that after I add multithreading support. Because, as I understand it, opencv already has this feature.

**Comment by jinfagang** [comment]
@Cykooz hoping for it!

**Comment by negezor** [comment]
It would be nice to add a comparison with libvips.

**Comment by vhbit** [comment]
I would also love to see comparison with `libvips` (not only raw performance, but also peak memory usage)

**Comment by bend-n** [comment]
With threads turned off, the results look like this:

![resizing](https://github.com/Cykooz/fast_image_resize/assets/70787919/29bd9661-beae-4a58-9539-cf272e137235)

benchmark code: https://github.com/bend-n/imenchmark/blob/main/benches/resizing.rs

**Comment by Cykooz** [comment]
@bend-n Thank you for your work.

**Comment by Cykooz** [comment]
@bend-n I have only one note about your opencv benchmark. OpenCV doesn't have Lanczos3 implementation. It has only Lanczos4. Lanczoz4 is more complicated (requires more calculations) than Lanczos3.

**Comment by bend-n** [comment]
Oh yeah i forgot to rename the field to `lanczos`.

**Comment by bend-n** [comment]
re: libvips, i cant seem to be able to get it to work properly.

**Comment by Cykooz** [comment]
I have found out that OpenCV returns incorrect results if we downscale image with different interpolations. All results, except INTER_AREA, look like INTER_NEAREST. With such results I can't do any correct benchmarks.

Image bellow is result of downscaling with different interpolations (result was upscaled x2 for better pixel visibility).
![OpenCV-resize](https://github.com/Cykooz/fast_image_resize/assets/163725/af3e574f-9376-47c7-808a-8c5635375b22)

**Comment by bend-n** [comment]
Why is that? Is opencv bugged?

**Comment by Cykooz** [comment]
Upscaling looks better. I don't know why downscaling is so bad.

**Comment by Cykooz** [comment]
I've understood the reason. OpenCV uses convolution with fixed kernal size. In this case, downscaling very big image into very small one looks like a result of nearest "interpolation".

fast_image_resize uses convolution with adaptive kernal size. It reuires more computations but makes more beter result.

It seems that only INTER_AREA in OpenCV uses something like adaptive kernal size.

**Comment by Cykooz** [comment]
@negezor I've done first comparison of `libvips` with `fast_image_resize`.
I've used `libvips` v8.12.1 from Ubuntu 22.04 repository.
`libvips` was tested in one-threaded mode with disabled caches.


|          | Nearest |  Box   | Bilinear | CatmullRom | Lanczos3 |
|----------|:-------:|:------:|:--------:|:----------:|:--------:|
| libvips  |  21.64  | 193.67 |  51.93   |   78.48    |  104.58  |
| fir rust |  0.86   | 61.49  |  97.43   |   165.89   |  232.39  |
| fir avx2 |    -    | 20.23  |  21.74   |   28.14    |  41.13   |

**Comment by PRadeckiDH** [comment]
Hi @Cykooz, are you still planning to add multithreading support?

**Comment by Cykooz** [comment]
Yes, I am planning it.

**Comment by bend-n** [comment]
whats the point of multithreading it? if you want it to be faster, write it on the gpu. opencv is way faster when you let it use a gpu

**Comment by Cykooz** [comment]
Our servers don't have GPU.
OpenCV has only one "interpolation" mode that is similar to convolution with adaptive kernal size - INTER_AREA (look previous comments).

**Comment by bend-n** [comment]
you can just scale multiple images at once, no need to scale one image in parallel?

**Comment by PRadeckiDH** [comment]
That can be a user decision to make, no? Test/use what makes the most sense in your use cases?

**Comment by Cykooz** [comment]
> you can just scale multiple images at once, no need to scale one image in parallel?

That was exactly the purpose of writing the crate a few years ago. I am using this crate to scale users images in web-service.
But now I want to add multi-threading support, which is more useful for single-user applications to work with single image at once (for example, image viewer application)

**Comment by Cykooz** [comment]
I released version 5.0.0. I added support for multi-thread image processing with the help of `rayon` crate.

**Comment by PRadeckiDH** [comment]
Awesome!! Will cykooz-resizer also receive multi-threaded support?

**Comment by Cykooz** [comment]
@PRadeckiDH Yes. I have plan to do it. I just need some time to figure out how to work with `rayon` thread-pool from python.
